Remove Sock::Release() and CloseSocket()

vasild opened the pull request:
_This is a piece of #21878, chopped off to ease review._

* `Sock::Release()` is unused, thus remove it
* `CloseSocket()` is only called from `Sock::Reset()`, so move the body of `CloseSocket()` inside `Sock::Reset()` and remove `CloseSocket()` - this helps to hide low level file descriptor sockets inside the `Sock` class.
* Rename `Sock::Reset()` to `Sock::Close()` and make it `private` - to be used only in the destructor and in the `Sock` assignment operator. This simplifies the public API by removing one method from it.

vasild committed:
net: remove CloseSocket()

Do the closing in `Sock::Reset()` and remove the standalone
`CloseSocket()`.

This reduces the exposure of low-level sockets (i.e. integer file
descriptors) outside of the `Sock` class.

vasild committed:
net: remove now unused Sock::Release()

laanwj commented on src/util/sock.cpp:
```diff
@@ -51,15 +51,21 @@ Sock& Sock::operator=(Sock&& other)
-SOCKET Sock::Release()
-{
-    const SOCKET s = m_socket;
+void Sock::Reset() {
```
`Reset` is a strange API function for a socket. Conceptually, I like `Close` better.

vasild replied:
Removed!

vasild committed:
net: rename Sock::Reset() to Sock::Close() and make it private

Outside of `Sock`, `Sock::Reset()` was used in just one place (in
`i2p.cpp`) which can use the assignment operator instead.

This simplifies the public `Sock` API by having one method less.

laanwj commented:
Concept ACK! Nice cleanup now!

laanwj commented on src/util/sock.cpp:
```diff
@@ -39,27 +39,28 @@ Sock::Sock(Sock&& other)
 }
 
-Sock::~Sock() { Reset(); }
+Sock::~Sock() { *this = Sock{INVALID_SOCKET}; }
```
Assigning *this in the descriptor looks a bit strange to me. Let's be sure that this isn't undefined behavior or footguns of some kind.

laanwj replied:
Oops! Yes, sounds like a more straightforwardly safe solution to me. Thanks.

vasild commented:
`e0537dada6...a724c39606`: make `Close()` `private` and call it from both the destructor and the assignment operator.

laanwj commented:
Code review ACK a724c39606273dfe4c6f9887ef8b77d0a98f1b34

laanwj commented on src/i2p.cpp:
```diff
@@ -410,7 +410,7 @@ void Session::Disconnect()
         }
     }
-    m_control_sock->Reset();
+    m_control_sock = std::make_unique<Sock>(INVALID_SOCKET);
```
Last nit: could this be `m_control_sock.release()`, or does the smart pointer always need to point to something?

vasild replied:
In the current code the smart pointer always has to point to something because we unconditionally dereference it in a few places.

However, that change is coming in https://github.com/bitcoin/bitcoin/pull/21878, from commit `i2p: avoid using Sock::Get() for checking for a valid socket` - it will use an empty unique_ptr to denote that there is no valid socket and together with a few other commits will remove the `Sock::Get()` method so that the value of the underlying socket file descriptor will be completely hidden inside the `Sock` class.

laanwj replied:
Great, so you already thought of that, thanks!
